src/Features/CSharp/Portable/CodeRefactorings/SyncNamespace/CSharpSyncNamespaceService.cs

```diff
@@ -0,0 +1,294 @@
+                    && firstMemberDeclaration.GetNameToken().Span.IntersectsWith(position);
+
+                shouldTrigger = shouldTrigger &&
+                    !(await ContainsPartialTypeWithMultipleDeclarationsAsync(document, compilationUnit, cancellationToken)
+                    .ConfigureAwait(false));
+
+                return (shouldTrigger, null);
+            }
+
+            return default;
+        }
+
+        /// <summary>
+        /// Try to change the namespace declaration based on the refacoring rules:
+        /// 
+        ///     - if neither declared and target namespace are "" (i.e. global namespace),
+        ///     then we try to change the name of the namespace.
+        ///     - if declared namespace is "", then we try to move all types declared 
+        ///     in global namespace in the document into a new namespace declaration.
+        ///     - if target namespace is "", then we try to move all members in declared 
+        ///     namespace to global namespace (i.e. remove the namespace declaration).    
+        /// </summary>
+        protected override SyntaxNode ChangeNamespaceDeclaration(
+            SyntaxNode root, 
+            ImmutableArray<string> declaredNamespaceParts, 
+            ImmutableArray<string> targetNamespaceParts)
+        {
+            if (declaredNamespaceParts.IsDefault || targetNamespaceParts.IsDefault || !(root is CompilationUnitSyntax compilationUnit))
+            {
+                return root;
+            }
+
+            // Move everything from global namespace to a namespace declaration
+            if (declaredNamespaceParts.Length == 1 && declaredNamespaceParts[0].Length == 0)
+            {
+                var targetNamespaceDecl = SyntaxFactory.NamespaceDeclaration(
+                    name: CreateNameSyntax(targetNamespaceParts, aliasQualifier: null, targetNamespaceParts.Length - 1)
+                    .WithAdditionalAnnotations(WarningAnnotation),
+                    externs: default, 
+                    usings: default,
+                    members: compilationUnit.Members);
+                return compilationUnit.WithMembers(new SyntaxList<MemberDeclarationSyntax>(targetNamespaceDecl))
+                    .WithAdditionalAnnotations(Formatter.Annotation);
+            }
+
+            // We should have a single member which is a namespace declaration in this compilation unit.
+            var namespaceDeclaration = root.DescendantNodes().OfType<NamespaceDeclarationSyntax>().Single();
+
+            // Move everything to global namespace
+            if (targetNamespaceParts.Length == 1 && targetNamespaceParts[0].Length == 0)
+            {
+                var triviaFromNamespaceDecl = GetOpeningAndClosingTriviaOfNamespaceDeclaration(namespaceDeclaration);
+
+                var members = namespaceDeclaration.Members;
+                var eofToken = compilationUnit.EndOfFileToken
+                    .WithAdditionalAnnotations(WarningAnnotation);
```

Comment:
nit: extraneous newline

Reply:
where is WarningAnnotation defined?  i didn't see it... also, why do we need a warning annotation on the EOF token?

Reply:
I just need to attach the warning to something, since in this case, the namespace declaration is removed, I figure EOF token would be a feasible choice.

Reply:
👍